### ISSUE TITLE
mount activity breaks ostree (any bind mounts in initramfs?)

### ISSUE THREAD

**Issue by cgwalters** [systemd/systemd#1556]
See: https://bugzilla.redhat.com/show_bug.cgi?id=1195761

Which links to: http://lists.freedesktop.org/archives/systemd-devel/2015-February/028482.html
and 
http://cgit.freedesktop.org/systemd/systemd/commit/?id=06e97888883e2cc12eb6514e80c7f0014295f59b

That in turn spawned a large thread on systemd-devel with several followups.  I think the current status is that systemd now handles activity for hardware-backed devices, but will still think that any bind mounts created in the initramfs are inactive.

If there's something OSTree could do better here I'm happy to make a change on that side.  Possibly generate unit files instead of just calling `mount`?  One problem I could see with doing that is that from systemd's perspective the paths change (`/mnt/sysroot/sysroot` is really `/sysroot`) etc.  It seems offhand it'd be simplest if we just preserved any mounts from the initramfs.

**Comment by poettering** [comment]
systemd puts devices that appear in /proc/self/mountinfo but where the backing device has not appeared yet in udev into a special "tentative" state, and does not consider them missing. If you see anything else, then this would be a bug. 

We only act on devices that showed up in both udev and /proc/self/mountinfo and then disappear from udev.

We explicitly support using "/bin/mount" as equivalent to writing unit files (modulo the deps that the latter follows, and the former of course does not) hence any misbehaviour in this area would be a bug...

**Comment by poettering** [comment]
Anyway, please explain exactly what the bug is, and what state you see which you don#t expect, and which you don't see that you do expect, and how we can reproduce this with a simple test case, otherwise this bug report is pretty useless.

**Comment by cgwalters** [comment]
`ostree-prepare-root.service`[1] is run in the initramfs and sets up a bind mount for `/` to `/sysroot` in the initramfs.  The symptom of the bug is we see:

```
systemd[1]: Unit sysroot.mount is bound to inactive service. Stopping, too.
```

I haven't traced through the code - since we're not unmounting `/var` which is also a bind mount, this might have to do with the fact that the `/sysroot` target is the rootfs mount?

[1] https://git.gnome.org/browse/ostree/tree/src/boot/ostree-prepare-root.service which executes
 https://git.gnome.org/browse/ostree/tree/src/switchroot/ostree-prepare-root.c

**Comment by poettering** [comment]
Well, if that's the log message you are seeing, then you are using an old version of systemd, from before February (418761f5864fe1ae7e9bb884259cc9933ae8ad0c). We do not track bugs in such old versions upstream. Please try to reproduce this with a current version, then reopen. Thanks.

To keep our workload small, please only report issues in the most recent two or three versions upstream. Thanks. Closing.

**Comment by cgwalters** [comment]
Sorry yes, that was an old log message.  You can reproduce from the images linked here:

https://getfedora.org/en/cloud/prerelease/atomic.html

```
curl -O https://download.fedoraproject.org/pub/fedora/linux/releases/test/23_Beta/Cloud/x86_64/Images/Fedora-Cloud-Atomic-23_Beta-20150915.x86_64.qcow2
```

The new message is:

```
sysroot.mount: Unit is bound to inactive unit dev-mapper-atomicos\x2droot.device. Stopping, too.
```

**Comment by cgwalters** [comment]
FWIW gnome-continuous is tracking systemd git master, and carrying a patch for this:

https://git.gnome.org/browse/gnome-continuous/tree/manifest.json#n105

**Comment by poettering** [comment]
Hmm, so if systemd thinks that sysroot.mount is bound to an inactive device "dev-mapper-atomicos\x2droot.device", then this suggests it is missing from /proc/self/mountinfo in this case. What precisely does /proc/self/mountinfo say about /sysroot when the issue happens?

When precisely do you get this message? During the deserialization phase of the initrd transition? Could you please provide a longer excerpt of the logs around the transition in debug mode (use systemd.log_level=debug on the kernel cmdline for that) if that's the case?

**Comment by cgwalters** [comment]
Journal looks like:

```
# journalctl b | grep '\(ostree\)\|\(sysroot\)
Oct 14 13:47:43 localhost kernel: Command line: BOOT_IMAGE=/ostree/fedora-atomic-115a3ec9198a518d0869d96390fc4fd240e28d8be774d0fb5823300e9312026c/vmlinuz-4.2.0-300.fc23.x86_64 no_timer_check console=tty1 console=ttyS0,115200n8 console=ttyS1 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.0/fedora-atomic/115a3ec9198a518d0869d96390fc4fd240e28d8be774d0fb5823300e9312026c/0
Oct 14 13:47:43 localhost kernel: Kernel command line: BOOT_IMAGE=/ostree/fedora-atomic-115a3ec9198a518d0869d96390fc4fd240e28d8be774d0fb5823300e9312026c/vmlinuz-4.2.0-300.fc23.x86_64 no_timer_check console=tty1 console=ttyS0,115200n8 console=ttyS1 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.0/fedora-atomic/115a3ec9198a518d0869d96390fc4fd240e28d8be774d0fb5823300e9312026c/0
Oct 14 13:47:43 localhost dracut-cmdline[160]: BOOT_IMAGE=/ostree/fedora-atomic-115a3ec9198a518d0869d96390fc4fd240e28d8be774d0fb5823300e9312026c/vmlinuz-4.2.0-300.fc23.x86_64 no_timer_check console=tty1 console=ttyS0,115200n8 console=ttyS1 rd.lvm.lv=atomicos/root root=/dev/mapper/atomicos-root ostree=/ostree/boot.0/fedora-atomic/115a3ec9198a518d0869d96390fc4fd240e28d8be774d0fb5823300e9312026c/0
Oct 14 13:47:44 localhost systemd[1]: Mounting /sysroot...
Oct 14 13:47:44 localhost systemd[1]: Mounted /sysroot.
Oct 14 13:47:45 localhost ostree-prepare-root[600]: Examining /sysroot//ostree/boot.0/fedora-atomic/115a3ec9198a518d0869d96390fc4fd240e28d8be774d0fb5823300e9312026c/0
Oct 14 13:47:45 localhost ostree-prepare-root[600]: Resolved OSTree target to: /sysroot/ostree/deploy/fedora-atomic/deploy/1d80c727e66083cff8f929ce42ff14aa13b6bd3a9fe0bb5cbfc705c0873e0d3f.0
Oct 14 13:47:45 localhost audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=ostree-prepare-root comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Oct 14 13:47:45 localhost audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=ostree-prepare-root comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Oct 14 13:47:46 localhost.localdomain systemd[1]: sysroot.mount: Unit is bound to inactive unit dev-mapper-atomicos\x2droot.device. Stopping, too.
Oct 14 13:47:46 localhost.localdomain systemd[1]: Unmounting /sysroot...
Oct 14 13:47:46 localhost.localdomain audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=ostree-remount comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Oct 14 13:47:46 localhost.localdomain audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=ostree-remount comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Oct 14 13:47:46 localhost.localdomain umount[651]: umount: /sysroot: not mounted
Oct 14 13:47:46 localhost.localdomain systemd[1]: sysroot.mount: Mount process exited, code=exited status=32
Oct 14 13:47:46 localhost.localdomain systemd[1]: Unmounted /sysroot.
Oct 14 13:47:46 localhost.localdomain systemd[1]: sysroot.mount: Unit entered failed state.
Oct 14 13:47:48 localhost.localdomain systemd[1]: sysroot.mount: Failed to run 'mount' task: No such file or directory
Oct 14 13:47:48 localhost.localdomain systemd[1]: Failed to mount /sysroot.
Oct 14 13:47:48 localhost.localdomain systemd[1]: Mounting /sysroot...
```

**Comment by cgwalters** [comment]
See also http://www.projectatomic.io/blog/2014/10/getting-started-with-cloud-init/ for information on how to configure the above linked cloud (qemu/KVM) image. You can log in using configured credentials after a bit of a timeout.

**Comment by poettering** [comment]
please provide the requested logs, thanks!

**Comment by cgwalters** [comment]
https://s3.amazonaws.com/cdn.verbum.org/journal-f23-atomic-host-beta-systemd.txt.gz

**Comment by poettering** [comment]
These logs suggest that the device unit in question stays "plugged" continously. Which makes this pretty weird, it's never inactive as the message suggests. 

Actually, we write debug messages on _every_ state change of device units, and it appears that some msgs got lost there, as we see the device being garbage collected without ever going into the "dead" state, which however is a requirement for that. 

I think there might be messages lost here, which is probably caused because the journal socket buffer is not large enough here. (background: we cannot allow that PID 1 deadlocks on the journal, which is a daemon it manages, hence we set a huge socket buffer and drop messages if it runs over.) Could you retry this stuff with systemd.log_target=kmsg and a huge kernel log buffer please?

222 is pretty old though, too. Any chance you can actually try this with something recent?

**Comment by dustymabe** [comment]
Here is some info [1] from a rawhide image [2] with systemd-227-1.fc24.x86_64 and kernel-4.3.0-0.rc4.git3.1.fc24.x86_64. I added the following to the kernel command line: `systemd.log_target=kmsg systemd.log_level=debug log_bug_len=1M`. 

[1] - http://paste.fedoraproject.org/279258/51683144/raw/
[2] - https://kojipkgs.fedoraproject.org//work/tasks/7730/11427730/Fedora-Cloud-Atomic-rawhide-20151013.x86_64.qcow2

UPDATE: In addition to the dmesg output in [1],here is the journal output [3]:
[3] - https://dustymabe.fedorapeople.org/journal.txt.gz

**Comment by dustymabe** [comment]
@poettering did the above output give you what you need?

**Comment by julianxhokaxhiu** [comment]
Up, this is breaking mdadm service too.

Now I'm not able anymore to boot my NAS because systemd v227 was pushed as an updated official core package on Arch.

**Comment by cgwalters** [comment]
Offhand, is there any reason to auto-unmount anything mounted in the initramfs?  If not then it seems like it'd be pretty easy to add a conditional.

**Comment by poettering** [comment]
That's not how the logic works: we have a dep tree, and that dep tree operates whenever a unit that is active is bound to a unit that is not active, and then pulls down the first unit. Whether the two units appeared during the initial initrd deserialization or later on doesn't really matter.

**Comment by dustymabe** [comment]
@poettering can you please answer the question in https://github.com/systemd/systemd/issues/1556#issuecomment-148748200

**Comment by poettering** [comment]
Sorry, but nope it didn't help. I am more puzzled than before. There appears to be a state transition for the units that isn't shown, but I don't see how that could ever get lost... I am seriously puzzled...

**Comment by msekletar** [comment]
I have a gut feeling that #2017 maybe solves this one also.

**Comment by dustymabe** [comment]
@msekletar thanks. Right now rawhide has `228-5.gite35a787.fc24`. I'll check it in a week or so once https://github.com/systemd/systemd/commit/97839ed6b7b19e2cb885a152b54fbd32334294e2 has made it into rawhide.

**Comment by msekletar** [comment]
@dustymabe I can prepare rawhide scratch build with just `97839ed` backported. Let me know in case you would like to test it earlier.

**Comment by dustymabe** [comment]
@msekletar thanks for the offer. I think it is ok though because I would have to build an ostree out of it and rebase to that. I'll just wait for it to get into rawhide and the rawhide ostree is updated so I can just use that. 

thanks again

**Comment by dustymabe** [comment]
@msekletar are you able to build a systemd package and actually push it into rawhide? If so then I'll take you up on your offer to build it :)

**Comment by msekletar** [comment]
Should land in rawhide today

**Comment by msekletar** [comment]
@dustymabe please test http://koji.fedoraproject.org/koji/buildinfo?buildID=704631

**Comment by dustymabe** [comment]
Thanks. I can't really test the rpm by itself without creating my own tree. Once it lands in rawhide and a new ostree compose picks it up I can test it out. Shouldn't be too long

**Comment by dustymabe** [comment]
Looks like this might just fix the problem. The rawhide ostree now boots without issue. I did the following to rebase from an f23 atomic host and the reboot worked just fine with no failed units on reboot. 

```
# setenforce 0
# ostree remote add rawhide --set=gpg-verify=false http://kojipkgs.fedoraproject.org/mash/atomic/rawhide/
# rpm-ostree rebase rawhide:fedora-atomic/rawhide/x86_64/docker-host
# reboot
```

@cgwalters do you mind confirming before we close this issue? @cgwalters would also be nice to know why I had to disable selinux to rebase to rawhide, but we can discuss that elsewhere.

**Comment by keszybz** [comment]
OK, let's assume that this is fixed by 97839ed6b7b19e2cb885a152b54fbd32334294e2.
